include/zephyr/device.h

```diff
@@ -1079,6 +1079,46 @@ __syscall int device_init(const struct device *dev);
 
 DT_FOREACH_STATUS_OKAY_NODE(Z_MAYBE_DEVICE_DECLARE_INTERNAL)
 
+/** @brief Expands to the full type. */
+#define Z_DEVICE_API_TYPE(_type) _type##_driver_api
+
+/**
+ * @brief Wrapper macro for declaring device API structs inside iterable sections.
+ *
+ * @param _type The device API type.
+ * @param _name The API instance name.
+ */
+#define DEVICE_API(_type, _name) const STRUCT_SECTION_ITERABLE(Z_DEVICE_API_TYPE(_type), _name)
+
+/**
+ * @brief Macro that evaluates to a boolean that can be used to check if an
+ *        API pointer is located in the iterable section.
+ *
+ * @param _type The device API type.
+ * @param _api The API instance pointer.
+ *
+ * @return true if the API pointer is located in the type's iterable section.
+ */
+#define DEVICE_API_IS(_type, _api)                                                                 \
+	({                                                                                         \
+		STRUCT_SECTION_START_EXTERN(Z_DEVICE_API_TYPE(_type));                             \
+		STRUCT_SECTION_END_EXTERN(Z_DEVICE_API_TYPE(_type));                               \
+		((const struct Z_DEVICE_API_TYPE(_type) *)_api >=                                  \
+			 STRUCT_SECTION_START(Z_DEVICE_API_TYPE(_type)) &&                         \
+		 (const struct Z_DEVICE_API_TYPE(_type) *)_api <                                   \
+			 STRUCT_SECTION_END(Z_DEVICE_API_TYPE(_type)));                            \
+	})
+
+/**
+ * @brief Macro that asserts if an API pointer isn't located in the type's iterable section.
+ *
+ * Note: The assertion depends on CONFIG_ASSERT
+ *
+ * @param _type The device API type.
+ * @param _api The API instance pointer.
+ */
+#define DEVICE_API_EVAL(_type, _api) __ASSERT_NO_MSG(DEVICE_API_IS(_type, _api))
```

Comment:
Another option could be to use `CHECKIF()` which allows the user to turn off these checks completely.